Sources/DeepLearning/Trainer.swift

```diff
@@ -0,0 +1,47 @@
+// Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// SupervisedTrainer implements a standard training loop for a model.
+public struct SupervisedTrainer<Model, Opt: Optimizer, Loss: Differentiable>
+    where Opt.Model == Model,
+          Loss: FloatingPoint,
+          Loss == Loss.CotangentVector {
+    // TODO: Convert to protocol to take advantage of argument labels to disambiguate.
```

Comment:
I don't think that's nice though. It kind of complicates the first-class concept just for argument labels, then we would need to add a bunch of conforming types that really just represent a function.

You can add argument disambiguation for function types here, FWIW:
```swift
@differentiable (_ predicted: Model.Output, _ expected: Model.Output) -> Loss
```

Reply:
I agree with you; I've removed the TODO (it'll show up the next time I push) and will resolve this conversation.